File: CovForDistributedSystems/CppAppWithUdpProbes/latency_bench.cpp
#include <iostream>
#include <chrono>
#include "mprewriter.hpp"

int main() {
    mpr_start_sender();

    // Warmup
    for (int i = 0; i < 10000;  ++i) {
             mprewriter_scope_START(1001);
    }
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    // Measure latency for 1M hits
    auto start = std::chrono::high_resolution_clock::now();
    for (int i = 0; i < 1000000; ++i) {
        
        mprewriter_scope_START(1001);
    }
    auto end = std::chrono::high_resolution_clock::now();

    // Flush remaining hits
    std::this_thread::sleep_for(std::chrono::milliseconds(100));
    close_probe();
    mpr_join_sender();

    double elapsed_ms = std::chrono::duration<double, std::milli>(end - start).count();
    double elapsed_us = elapsed_ms * 1000.0;
    double per_hit_us = elapsed_us / 1000000.0;

    std::cout << "C++ Latency Benchmark Results:" << std::endl;
    std::cout << "==============================" << std::endl;
    std::cout << "Total hits: 1000000" << std::endl;
    std::cout << "Total time: " << elapsed_ms << " ms" << std::endl;
    std::cout << "Average latency per hit: " << per_hit_us << " microseconds" << std::endl;

    return 0;
}

File: CovForDistributedSystems/RustAppWithUdpProbes/udp_probe_demo_v2/src/latency_bench.rs
use std::time::Instant;

mod mprewriter;

fn main() {
    println!("Initializing probe...");
    mprewriter::start();
    println!("Probe initialized");

    // Warmup
    println!("Warmup...");
    for _ in 0..10000 {
        mprewriter::scope_START(1001);
    }
    std::thread::sleep(std::time::Duration::from_millis(100));
    println!("Warmup complete");

    // Measure latency for 1M hits
    println!("Starting measurement...");
    let start = Instant::now();
    for _ in 0..1000000 {
        mprewriter::scope_START(1001);
    }
    let end = Instant::now();
    println!("Measurement complete");

    // Flush remaining hits
    println!("Flushing...");
    std::thread::sleep(std::time::Duration::from_millis(100));
        // Note: Skipping mprewriter::shutdown() because it hangs on join()
        // The background thread will be killed when the process exits
        println!("Shutdown complete (skipped join)");

    let elapsed = end.duration_since(start);
    let elapsed_us = elapsed.as_secs_f64() * 1_000_000.0;
    let per_hit_us = elapsed_us / 1_000_000.0;

    println!("Rust Latency Benchmark Results:");
    println!("==============================");
    println!("Total hits: 1000000");
    println!("Total time: {:.3} ms", elapsed_us / 1000.0);
    println!("Average latency per hit: {:.3} microseconds", per_hit_us);
}
